### ISSUE TITLE
Initial upload (after app start) results in "Your socket connection to the server was not read from or written to within the timeout period. Idle connections will be closed"

### ISSUE THREAD

**Issue by caleb0199** [aws/aws-sdk-js#1647]
I've got a node app that uploads files to s3. 
The app runs in a docker container.

It's not a new app (was created almost a year ago). 

An issue started happening with aws-sdk version 2.68 (did not happen in prior versions).
Changelog indicates a change in default signature version in v2.68, so I've tried specifying signatureVersion: 'v2' in a config.update() call, but that doesn't seem to make any difference. Not sure if the issue is related to that change or not, but nothing else is documented regarding S3 changes in v2.68.

Here's the problem:
The first upload (after container/app start) fails after a timeout with the message: "Your socket connection to the server was not read from or written to within the timeout period. Idle connections will be closed."
All subsequent uploads succeed. (Also, the initial download succeeds, as do all subsequent ones, so the issue seems to be only with the initial upload.)

Stack trace accompanying the error is appended below.
Any help figuring out cause and fix would be appreciated.

Stack trace:
```
"RequestTimeout: Your socket connection to the server was not read from or written to within the timeout period. Idle connections will be closed.\n
at Request.extractError (/usr/src/app/node_modules/aws-sdk/lib/services/s3.js:577:35)\n
at Request.callListeners (/usr/src/app/node_modules/aws-sdk/lib/sequential_executor.js:105:20)\n
at Request.emit (/usr/src/app/node_modules/aws-sdk/lib/sequential_executor.js:77:10)\n
at Request.emit (/usr/src/app/node_modules/aws-sdk/lib/request.js:683:14)\n
at Request.transition (/usr/src/app/node_modules/aws-sdk/lib/request.js:22:10)\n
at AcceptorStateMachine.runTo (/usr/src/app/node_modules/aws-sdk/lib/state_machine.js:14:12)\n
at /usr/src/app/node_modules/aws-sdk/lib/state_machine.js:26:10\n
at Request.<anonymous> (/usr/src/app/node_modules/aws-sdk/lib/request.js:38:9)\n
at Request.<anonymous> (/usr/src/app/node_modules/aws-sdk/lib/request.js:685:12)\n
at Request.callListeners (/usr/src/app/node_modules/aws-sdk/lib/sequential_executor.js:115:18)\n
at Request.emit (/usr/src/app/node_modules/aws-sdk/lib/sequential_executor.js:77:10)\n
at Request.emit (/usr/src/app/node_modules/aws-sdk/lib/request.js:683:14)\n
at Request.transition (/usr/src/app/node_modules/aws-sdk/lib/request.js:22:10)\n
at AcceptorStateMachine.runTo (/usr/src/app/node_modules/aws-sdk/lib/state_machine.js:14:12)\n
at /usr/src/app/node_modules/aws-sdk/lib/state_machine.js:26:10\n
at Request.<anonymous> (/usr/src/app/node_modules/aws-sdk/lib/request.js:38:9)\n
at Request.<anonymous> (/usr/src/app/node_modules/aws-sdk/lib/request.js:685:12)\n
at Request.callListeners (/usr/src/app/node_modules/aws-sdk/lib/sequential_executor.js:115:18)\n
at callNextListener (/usr/src/app/node_modules/aws-sdk/lib/sequential_executor.js:95:12)\n
at IncomingMessage.onEnd (/usr/src/app/node_modules/aws-sdk/lib/event_listeners.js:269:13)\n
at emitNone (events.js:91:20)\n
at IncomingMessage.emit (events.js:185:7)\n
at endReadableNT (_stream_readable.js:974:12)\n
at _combinedTickCallback (internal/process/next_tick.js:80:11)\n
at process._tickDomainCallback (internal/process/next_tick.js:128:9)","code":"RequestTimeout","region":null,"name":"RequestTimeout","time":"2017-07-25T19:53:32.349Z","requestId":"CCBACEC5B4B1A0D0","extendedRequestId":"48ZcbTpnCysSpv5kumeGqAoYZ0DKzYf2AYL2Kv/Y1FVrNJl1MnImsBS+Y9D3xLqVDrqFRzhDVw+MBLgfPQXgDfzeOTSfjZR0","statusCode":400,"retryable":true}
```

**Comment by jeskew** [comment]
Hi @caleb0199,

In which region is the S3 bucket located, and are you providing that region as configuration to your S3 client?

**Comment by caleb0199** [comment]
First, thanks so much for your help, @jeskew 

We're using region: 'us-east-1'
When I output the s3 client instance to the log immediately before the call to s3.putObject(), it shows that value.
so, in our code, I have:
```
var s3 = new AWS.S3({params: params})
  console.log('aws.s3:', s3)
  s3.putObject({ContentLength: req.headers['content-length'], Body: req}, function (err) {
    callback(err, remote_source, remote_file)
  })
```
and the output includes:
```
  | aws.s3: Service {
  |   config:
  |    Config {
  |      credentials:
  |       Credentials {
  |         expired: false,
  |         expireTime: null,
  |         accessKeyId: removed for security
  |         sessionToken: undefined },
  |      credentialProvider: CredentialProviderChain { providers: [Object] },
  |      region: 'us-east-1',
  |      logger: null,
  |      apiVersions: {},
  |      apiVersion: '2016-08-09',
  |      endpoint: 's3.amazonaws.com',
  |      httpOptions: { timeout: 120000 },
  |      maxRetries: undefined,
  |      maxRedirects: 10,
  |      paramValidation: true,
  |      sslEnabled: true,
  |      s3ForcePathStyle: false,
  |      s3BucketEndpoint: false,
  |      s3DisableBodySigning: true,
  |      computeChecksums: true,
  |      convertResponseTypes: true,
  |      correctClockSkew: false,
  |      customUserAgent: null,
  |      dynamoDbCrc32: true,
  |      systemClockOffset: 0,
  |      signatureVersion: 'v2',
  |      signatureCache: true,
  |      retryDelayOptions: {},
  |      useAccelerateEndpoint: false,
  |      params: { Bucket: 'our_bucket_name', Key: 'our_file_name' } },
  |   isGlobalEndpoint: false,
  |   endpoint:
  |    Endpoint {
  |      protocol: 'https:',
  |      host: 's3.amazonaws.com',
  |      port: 443,
  |      hostname: 's3.amazonaws.com',
  |      pathname: '/',
  |      path: '/',
  |      href: 'https://s3.amazonaws.com/' },
  |   _clientId: 1 
}
```
The output of the s3 client instance is the same for the second call, except the last attribute, "_clientId", is 2, instead of 1.

Thanks again for your help!

**Comment by caleb0199** [comment]
Any ideas on this one? It's still marked "Feedback Requested"...not sure if I should/can somehow clear that label after having responded (above) to the request.

**Comment by jeskew** [comment]
If you restart the application in an already running container, do you see the same issue?

**Comment by caleb0199** [comment]
I've never tried that. Trying to figure out how to do that. I'm in the terminal of the container, but killing any of the app processes causes the container to stop. Still experimenting.

**Comment by caleb0199** [comment]
I haven't figured out how to do that, yet. Is that info key to diagnosing?

**Comment by jeskew** [comment]
If the upload is timing out because network or filesystem access is considerably slower on a fresh container, there might be a workaround in docker for that scenario.

**Comment by caleb0199** [comment]
In order to mitigate impact to our app, I'm going to back up to v2.67.0 (which doesn't exhibit this behavior).
After that, I'll try to figure out how to restart app without re-starting container. But, as a workaround for me, reverting to 2.27 is easier.
Strange that it only impacts upload (as first operation). If I download as my first operation after restart, that works fine. And a subsequent upload also works. But, if upload is my first operation, that fails the first time, and then succeeds. 
I'd like to get to root cause on this. Seems like others are not complaining about this, so it may be something I'm doing as config...but not sure why the change in behavior between 2.67 and 2.68.

**Comment by tvald** [comment]
I'm also seeing this problem. When submitted in sequence, the first `putObject()` fails and subsequent `putObject()`s succeed. When submitted in parallel, all `putObject()`s fail.

Using `aws-sdk` version `2.94.0`.

**Comment by chux0519** [comment]
the same issue

**Comment by srchase** [comment]
@caleb0199 

Are you seeing this issue with the current version of the SDK?

**Comment by caleb0199** [comment]
I'm sorry I can't be of help. I no longer work at the company where that
project exists, so I don't know if it's still happening with the latest
version of the SDK.

**Comment by srchase** [comment]
I'm going to close this issue then.  If anyone has this issue with a reliable repro, feel free to open a new issue.
